File: daemon/selfinstall.js
/**
 * Self-Install Bootstrap
 *
 * When the user runs the minimal install command:
 *
 *   docker run -v /var/run/docker.sock:/var/run/docker.sock ghcr.io/besoeasy/yantr
 *
 * This module detects that the container is NOT running with the full required
 * configuration (host network + volumes mount) and uses the Docker socket to:
 *
 *   1. Pull the latest image (if not already present)
 *   2. Remove any existing "yantr" container
 *   3. Launch a new "yantr" container with the full production configuration
 *   4. Exit this bootstrap container
 *
 * If the container is already running with the correct configuration (i.e. the
 * user ran the full command, or this is the properly-installed container), this
 * module is a no-op and normal startup continues.
 */

import Docker from "dockerode";
import os from "os";

const CONTAINER_NAME = process.env.YANTR_CONTAINER_NAME || "yantr";
const IMAGE = process.env.YANTR_IMAGE || "ghcr.io/besoeasy/yantr";
const SOCKET_PATH = process.env.DOCKER_SOCKET || "/var/run/docker.sock";

/**
 * Check whether we are already running as the fully-configured container.
 *
 * We consider the current container "fully configured" when ALL of the
 * following are true:
 *   - It is named CONTAINER_NAME (i.e. it was started by a previous install run)
 *   - It has NetworkMode === "host"
 *   - It has /var/lib/docker/volumes bind-mounted
 *
 * If any of these are missing we are the bootstrap (minimal) container and
 * should re-launch ourselves properly.
 */
async function isFullyConfigured(docker) {
  // Resolve our own container ID from the cgroup / hostname
  const hostname = os.hostname();

  let selfInfo = null;
  try {
    // In Docker, the container short-id is the hostname by default
    const container = docker.getContainer(hostname);
    selfInfo = await container.inspect();
  } catch {
    // Could not inspect by hostname — not inside a container at all (dev mode)
    return true;
  }

  const hostConfig = selfInfo?.HostConfig || {};
  const name = (selfInfo?.Name || "").replace(/^\//, "");

  const hasHostNetwork = hostConfig.NetworkMode === "host";
  const binds = hostConfig.Binds || [];
  const hasVolumeMount = binds.some((b) => b.includes("/var/lib/docker/volumes"));
  const isNamedCorrectly = name === CONTAINER_NAME;

  return hasHostNetwork && hasVolumeMount && isNamedCorrectly;
}

/**
 * Remove a container by name if it exists (stop it first if running).
 */
async function removeExisting(docker, name) {
  try {
    const container = docker.getContainer(name);
    const info = await container.inspect();
    if (info.State?.Running) {
      console.log(`[selfinstall] Stopping existing "${name}" container...`);
      await container.stop({ t: 5 });
    }
    console.log(`[selfinstall] Removing existing "${name}" container...`);
    await container.remove({ force: true });
  } catch (err) {
    if (err?.statusCode === 404) return; // doesn't exist, nothing to do
    throw err;
  }
}

/**
 * Run the self-install bootstrap.
 * Returns true if the bootstrap ran (caller should exit immediately after).
 * Returns false if we are already fully configured (normal startup continues).
 */
export async function runSelfInstallIfNeeded() {
  const docker = new Docker({ socketPath: SOCKET_PATH });

  let fullyConfigured = false;
  try {
    fullyConfigured = await isFullyConfigured(docker);
  } catch (err) {
    console.warn(`[selfinstall] Could not check configuration: ${err.message}`);
    return false;
  }

  if (fullyConfigured) {
    return false; // already correct — proceed with normal startup
  }

  console.log("[selfinstall] Minimal install detected. Launching fully-configured yantr container...");

  // Remove any existing container with this name
  await removeExisting(docker, CONTAINER_NAME);

  // Create and start the new container with full production config
  const container = await docker.createContainer({
    name: CONTAINER_NAME,
    Image: IMAGE,
    HostConfig: {
      NetworkMode: "host",
      RestartPolicy: { Name: "unless-stopped" },
      Binds: [
        `${SOCKET_PATH}:/var/run/docker.sock`,
        "/var/lib/docker/volumes:/var/lib/docker/volumes",
      ],
    },
    Env: buildEnv(),
  });

  await container.start();

  console.log(`[selfinstall] yantr container started successfully.`);
  console.log(`[selfinstall] Access the UI at http://localhost:5252`);
  console.log(`[selfinstall] This bootstrap container will now exit.`);

  return true; // signal to caller: exit
}

/**
 * Forward whitelisted env vars from the current process to the new container.
 * This lets users pass TZ, YANTR_* vars through the minimal install command.
 */
function buildEnv() {
  const passthrough = [
    "TZ",
    "YANTR_CONTAINER_NAME",
    "YANTR_IMAGE",
    "YANTR_SELFUPDATE",
    "YANTR_SELFUPDATE_INTERVAL",
    "DOCKER_SOCKET",
    "UI_BASE_PATH",
    "VITE_BASE_PATH",
  ];

  const env = [];
  for (const key of passthrough) {
    if (process.env[key] !== undefined) {
      env.push(`${key}=${process.env[key]}`);
    }
  }
  return env;
}

File: core/selfinstall/selfinstall.go
// Package selfinstall implements the bootstrap self-install logic.
//
// When the user runs the minimal install:
//
//	docker run -v /var/run/docker.sock:/var/run/docker.sock ghcr.io/besoeasy/yantr
//
// This detects the container is NOT fully configured (host network + volumes mount),
// then uses the Docker socket to:
//  1. Remove any existing "yantr" container
//  2. Launch a new container with full production configuration
//  3. Exit this bootstrap container
package selfinstall

import (
	"context"
	"fmt"
	"os"
	"strings"

	dockertypes "github.com/docker/docker/api/types/container"
	"github.com/docker/docker/api/types/network"
	"core/docker"
)

var (
	containerName = envOr("YANTR_CONTAINER_NAME", "yantr")
	imageName     = envOr("YANTR_IMAGE", "ghcr.io/besoeasy/yantr")
	socketPath    = envOr("DOCKER_SOCKET", "/var/run/docker.sock")
)

func envOr(key, def string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return def
}

// RunIfNeeded checks whether we are fully configured. Returns true if the
// bootstrap ran (caller should exit); false if normal startup should continue.
func RunIfNeeded() (bool, error) {
	fully, err := isFullyConfigured()
	if err != nil {
		fmt.Printf("[selfinstall] Could not check configuration: %v\n", err)
		return false, nil
	}
	if fully {
		return false, nil
	}

	fmt.Println("[selfinstall] Minimal install detected. Launching fully-configured yantr container...")

	if err := removeExisting(containerName); err != nil {
		return false, fmt.Errorf("failed to remove existing container: %w", err)
	}

	if err := launchFullContainer(); err != nil {
		return false, fmt.Errorf("failed to launch container: %w", err)
	}

	fmt.Println("[selfinstall] yantr container started successfully.")
	fmt.Println("[selfinstall] Access the UI at http://localhost:5252")
	fmt.Println("[selfinstall] This bootstrap container will now exit.")
	return true, nil
}

func isFullyConfigured() (bool, error) {
	hostname, err := os.Hostname()
	if err != nil {
		return true, nil
	}

	info, err := docker.ContainerInspect(context.Background(), hostname)
	if err != nil {
		// Not inside a container (dev mode) — treat as fully configured
		return true, nil
	}

	hc := info.HostConfig
	name := strings.TrimPrefix(info.Name, "/")

	hasHostNetwork := hc.NetworkMode == "host"
	hasVolMount := false
	for _, bind := range hc.Binds {
		if strings.Contains(bind, "/var/lib/docker/volumes") {
			hasVolMount = true
			break
		}
	}
	isNamed := name == containerName

	return hasHostNetwork && hasVolMount && isNamed, nil
}

func removeExisting(name string) error {
	info, err := docker.ContainerInspect(context.Background(), name)
	if err != nil {
		return nil // doesn't exist
	}
	if info.State.Running {
		fmt.Printf("[selfinstall] Stopping existing %q container...\n", name)
		_ = docker.ContainerStop(context.Background(), name, dockertypes.StopOptions{Timeout: intPtr(5)})
	}
	fmt.Printf("[selfinstall] Removing existing %q container...\n", name)
	return docker.ContainerRemove(context.Background(), name, dockertypes.RemoveOptions{Force: true})
}

func launchFullContainer() error {
	config := &dockertypes.Config{
		Image: imageName,
		Env:   buildEnv(),
	}
	hostConfig := &dockertypes.HostConfig{
		NetworkMode:   "host",
		RestartPolicy: dockertypes.RestartPolicy{Name: "unless-stopped"},
		Binds: []string{
			socketPath + ":/var/run/docker.sock",
			"/var/lib/docker/volumes:/var/lib/docker/volumes",
		},
	}
	networkConfig := &network.NetworkingConfig{}

	resp, err := docker.ContainerCreate(
		context.Background(),
		config,
		hostConfig,
		networkConfig,
		nil,
		containerName,
	)
	if err != nil {
		return err
	}
	return docker.ContainerStart(context.Background(), resp.ID, dockertypes.StartOptions{})
}

var passThroughEnvKeys = []string{
	"TZ", "YANTR_CONTAINER_NAME", "YANTR_IMAGE",
	"YANTR_AUTH_SECRET", "YANTR_AUTH_USERNAME",
	"YANTR_SELFUPDATE", "YANTR_SELFUPDATE_INTERVAL",
	"DOCKER_SOCKET", "UI_BASE_PATH", "VITE_BASE_PATH",
}

func buildEnv() []string {
	var env []string
	for _, k := range passThroughEnvKeys {
		if v := os.Getenv(k); v != "" {
			env = append(env, k+"="+v)
		}
	}
	return env
}

func intPtr(n int) *int { return &n }
